### DISCUSSION TITLE
Unable to retrieve Environment variables in Unit Test for Cake extension

### DISCUSSION THREAD

**Discussion by nick5454** [cake-build/cake#3806] [General]
I'm running a Unit Test for a cake extension. And i'm trying to retrieve a local environment variable, but it's not working. Do I need to just fake the variable and this will work when ran in the build pipeline?

This returns null

var home = context.Environment.GetEnvironmentVariable("JAVA_HOME");

**Comment by devlead** [comment]
Hard to answer without seeing code, but generally when testing something like Cake.Testing / NSubstitute is used.

NSubstitute example here
https://github.com/cake-build/cake/blob/d454dc2a282879c922be8e4a195583f0357b3dda/src/Cake.Common.Tests/Fixtures/Build/GitHubActionsInfoFixture.cs#L19-L26

**Reply by nick5454** [reply]
@devlead Thanks! I'm testing the ENV vars to know whether Java exists on the machine. If it doesn't then I need to download and install. And set the Env var JAVA_HOME. Can you suggest the best way to achieve this? 

I'm assuming since I have the download url. I can download it and then unzip it and install via silent mode. Or is there a better way?

**Reply by devlead** [reply]
`JAVA_HOME` is one way, another way would be to try to resolve java executable.

I.e. something like

`Context.Tools.Resolve("java.exe") ?? Context.Tools.Resolve("java")`

Which is the way most Cake tool resolution is done by default i.e.

https://github.com/cake-build/cake/blob/d454dc2a282879c922be8e4a195583f0357b3dda/src/Cake.Common/Tools/DotNet/DotNetTool.cs#L51

**Comment by nick5454** [comment]
@devlead  Will that automatically download the Java executable if it doesn't exist on the machine?

The issue I have is that in order for the SDKManager with Android to work, the JAVA_HOME environment variable has to be pointed to the folder where it was installed. But on a clean Release MS Hosted pipeline does not have android or Java installed

My basic path is 
Download then install the JDK
create JAVA_HOME and point it to the correct folder
Download Android Command Tools to a folder ( JAVA_HOME must be pointing to the correct folder )
run the tool and install the needed SDK's

Is there an example of doing that? I'm looking at Cake.AndroidSDKManager and he's using an http call to grab the installation file and executing.

I assumed that's what I should do or is this wrong?
